Fix coding style to avoid unnecessary computation and copying objects

canhld94 opened the pull request:
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

canhld94 committed:
fix coding style to avoid unecessary copy objects

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=34565) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=34565) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=34565) it.</sub>

canhld94 committed:
fix style check

canhld94 commented:
Failed tests look like unrelated.

amosbird commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -978,18 +981,16 @@ MergeTreeDataSelectAnalysisResultPtr ReadFromMergeTree::selectRangesToRead(
 }
 
-ReadFromMergeTree::AnalysisResult ReadFromMergeTree::getAnalysisResult() const
+void ReadFromMergeTree::checkAnalysisResult() const
```
Let's keep the `getAnalysisResult()` but returning reference instead. We can build two variants:

```
    AnalysisResult & getAnalysisResult();
    const AnalysisResult & getAnalysisResult() const
    {
        return const_cast<ReadFromMergeTree *>(this)->getAnalysisResult();
    }
```

amosbird commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -1211,8 +1214,9 @@ void ReadFromMergeTree::describeActions(JSONBuilder::JSONMap & map) const
-    auto index_stats = std::move(result.index_stats);
+    checkAnalysisResult();
+    const auto & result = std::get<ReadFromMergeTree::AnalysisResult>(analyzed_result_ptr->result);
+    const auto & index_stats = result.index_stats;
```
`const auto & index_stats = getAnalysisResult().index_stats`

canhld94 committed:
keep getAnalysisResult and return ref

canhld94 commented:
Hi @amosbird @evillique I addressed all comments. Failed test case looks unrelated. The performance comparision is somehow unstable, but I think the logic improvement is quite clear.

evillique commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -365,7 +369,7 @@ static ActionsDAGPtr createProjection(const Block & header)
 
 Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsWithOrder(
-    RangesInDataParts && parts_with_ranges,
+    RangesInDataParts parts_with_ranges,
```
So it seems like before with `&&` we didn't need to call move constructor (and additional destructor at the end) to pass and take elements from `parts_with_ranges` and now we explicitly do. Am I missing something or how is this part better than before?

alexey-milovidov commented:
@canhld94 please resolve conflicts.

canhld94 commented:
> @canhld94 please resolve conflicts.

Done

alexey-milovidov commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -1014,11 +1017,15 @@ void ReadFromMergeTree::setQueryInfoInputOrderInfo(InputOrderInfoPtr order_info)
+
+const ReadFromMergeTree::AnalysisResult & ReadFromMergeTree::getAnalysisResult() const
+{
+    return const_cast<ReadFromMergeTree *>(this)->getAnalysisResult();
```
This makes code less readable.

amosbird replied:
That's the only legitimate use of const_cast as far as I know, to avoid code duplication when overriding const/non-const methods. Similar examples also exist in book C++ Primer.

canhld94 replied:
Looks like both versions are const method now, but one return value and one return const ref. The 1st version is a mistake, should be non-const and return `AnalysisResult &`

canhld94 replied:
I found that the non-const version of this method currently not used, so I just remove it @alexey-milovidov

alexey-milovidov reviewed (changes requested):
Let's remove const_cast.

canhld94 committed:
Merge branch 'master' into ch_canhld_fix

canhld94 committed:
Remove non-const version of getAnalysisResult

canhld94 committed:
Trigger Build

canhld94 committed:
Merge branch 'master' into ch_canhld_fix

alexey-milovidov committed:
Merge branch 'master' into ch_canhld_fix

alexey-milovidov commented:
`clang-tidy` pointed out that the code is non-perfect:

```
Aug 02 12:10:49 /build/build_docker/../src/Processors/QueryPlan/ReadFromMergeTree.cpp:1021:32: error: passing result of std::move() as a const reference argument; no move will actually happen [performance-move-const-arg,-warnings-as-errors]
Aug 02 12:10:49         std::rethrow_exception(std::move(std::get<std::exception_ptr>(analyzed_result_ptr->result)));
Aug 02 12:10:49                                ^~~~~~~~~~                                                         ~
```

alexey-milovidov committed:
Update ReadFromMergeTree.cpp

alexey-milovidov commented:
Please reopen when ready.
